mitiq/adaptive_zne.py

```diff
@@ -0,0 +1,100 @@
+from pyquil import Program
+from typing import List, Tuple, Callable
+
+from mitiq.zne import richardson_extr
+import mitiq.qiskit.qiskit_utils as qs_utils
+from mitiq import QPROGRAM
+
+
+class Factory(object):
+    """
+    This object adaptively produces new noise scaling parameters based on a historical stack of previous noise
+    scale parameters and previously calculated expectation values. Noise scaling parameters are stored on the `in_stack`
+    and the running list of expectation values are stored on the `out_stack`.
+    """
+    def __init__(self, instack: List[float], outstack: List[float]) -> None:
+        self.instack = instack
+        self.outstack = outstack
+
+    def step(self, instack: List[float], outstack: List[float]) -> None:
+        raise NotImplementedError
+
+    def is_converged(self, instack: List[float], outstack: List[float]) -> None:
+        raise NotImplementedError
+
+    def reduce(self, expectations: List[float]) -> float:
```

Comment:
Going with @andreamari's comment, maybe `reduce` should have an argument that specifies the method?